In [1]:
import pandas as pd

1. Attendance Data

    Load the attendance.csv file and calculate an attendnace percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

    You should end up with something like this:


name

Billy    0.5250

Jane     0.6875

John     0.9125

Sally    0.7625

Name: grade, dtype: float64

In [95]:
attendance = pd.read_csv("attendance.csv")

In [96]:
attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [97]:
attendance_melt = attendance.melt(id_vars="Unnamed: 0", var_name="date", value_name="attendance")

In [98]:
attendance_melt.attendance = attendance_melt.attendance.str.replace("P","1")

In [99]:
attendance_melt.attendance = attendance_melt.attendance.str.replace("H","0.5")

In [100]:
attendance_melt.attendance = attendance_melt.attendance.str.replace("A","-1")

In [101]:
attendance_melt.attendance = attendance_melt.attendance.str.replace("T",".9")

In [102]:
attendance_melt.attendance = attendance_melt.attendance.astype("float")

In [103]:
attendance_melt = attendance_melt.rename(columns={"Unnamed: 0": "name"})

In [104]:
attendance_melt

name        date  attendance
0   Sally  2018-01-01         1.0
1    Jane  2018-01-01        -1.0
2   Billy  2018-01-01        -1.0
3    John  2018-01-01         1.0
4   Sally  2018-01-02         0.9
5    Jane  2018-01-02         1.0
6   Billy  2018-01-02         0.9
7    John  2018-01-02         0.9
8   Sally  2018-01-03         0.9
9    Jane  2018-01-03         0.9
10  Billy  2018-01-03        -1.0
11   John  2018-01-03         0.5
12  Sally  2018-01-04         0.5
13   Jane  2018-01-04         0.9
14  Billy  2018-01-04        -1.0
15   John  2018-01-04         1.0
16  Sally  2018-01-05         1.0
17   Jane  2018-01-05         0.9
18  Billy  2018-01-05         0.5
19   John  2018-01-05         1.0
20  Sally  2018-01-06        -1.0
21   Jane  2018-01-06         0.9
22  Billy  2018-01-06         0.9
23   John  2018-01-06         0.9
24  Sally  2018-01-07         0.9
25   Jane  2018-01-07        -1.0
26  Billy  2018-01-07         1.0
27   John  2018-01-07         1.0
28  Sally  2018-01-08         0.9
29   Jane  2018-01-08         0.9
30  Billy  2018-01-08         0.9
31   John  2018-01-08         1.0

In [105]:
attendance_melt.groupby("name").mean()

attendance
name             
Billy      0.1500
Jane       0.4375
John       0.9125
Sally      0.6375

2. Coffee Levels

    Read the coffee_levels.csv file.
    Transform the data so that each carafe is in it's own column.
    Is this the best shape for the data?

In [106]:
coffee = pd.read_csv("coffee_levels.csv")

In [107]:
coffee

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

In [137]:
coffee_df = coffee.pivot_table(values="coffee_amount", index="hour", columns="coffee_carafe").reset_index()

In [141]:
coffee_df.columns.name = ""

In [142]:
coffee_df

hour         x         y         z
0     8  0.816164  0.189297  0.999264
1     9  0.451018  0.521502  0.915990
2    10  0.843279  0.023163  0.144928
3    11  0.335533  0.235529  0.311495
4    12  0.898291  0.017009  0.771947
5    13  0.310711  0.997464  0.398520
6    14  0.507288  0.058361  0.864464
7    15  0.215043  0.144644  0.436364
8    16  0.183891  0.544676  0.280621
9    17  0.391560  0.594126  0.436677

3. Cake Recipes

    a. Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.


In [109]:
cakes = pd.read_csv("cake_recipes.csv")

In [110]:
cakes

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

b. Tidy the data as necessary.

In [111]:
recipe_pos = cakes["recipe:position"].str.split(":", expand=True)

In [112]:
recipe_pos.columns = ["recipe", "postition"]

In [113]:
recipe_pos

recipe postition
0      a    bottom
1      a       top
2      b    bottom
3      b       top
4      c    bottom
5      c       top
6      d    bottom
7      d       top

In [116]:
cakes = pd.concat([recipe_pos, cakes], axis=1).drop(columns="recipe:position")

In [117]:
cakes

recipe postition        225        250        275        300
0      a    bottom  61.738655  53.912627  74.414730  98.786784
1      a       top  51.709751  52.009735  68.576858  50.228470
2      b    bottom  57.095320  61.904369  61.196980  99.248541
3      b       top  82.455004  95.224151  98.594881  58.169349
4      c    bottom  96.470207  52.001358  92.893227  65.473084
5      c       top  71.306308  82.795477  92.098049  53.960273
6      d    bottom  52.799753  58.670419  51.747686  56.183110
7      d       top  96.873178  76.101363  59.571620  50.971626

In [119]:
cake_melt = cakes.melt(id_vars=["recipe", "postition"], var_name="temp", value_name="score")

In [120]:
cake_melt

recipe postition temp      score
0       a    bottom  225  61.738655
1       a       top  225  51.709751
2       b    bottom  225  57.095320
3       b       top  225  82.455004
4       c    bottom  225  96.470207
5       c       top  225  71.306308
6       d    bottom  225  52.799753
7       d       top  225  96.873178
8       a    bottom  250  53.912627
9       a       top  250  52.009735
10      b    bottom  250  61.904369
11      b       top  250  95.224151
12      c    bottom  250  52.001358
13      c       top  250  82.795477
14      d    bottom  250  58.670419
15      d       top  250  76.101363
16      a    bottom  275  74.414730
17      a       top  275  68.576858
18      b    bottom  275  61.196980
19      b       top  275  98.594881
20      c    bottom  275  92.893227
21      c       top  275  92.098049
22      d    bottom  275  51.747686
23      d       top  275  59.571620
24      a    bottom  300  98.786784
25      a       top  300  50.228470
26      b    bottom  300  99.248541
27      b       top  300  58.169349
28      c    bottom  300  65.473084
29      c       top  300  53.960273
30      d    bottom  300  56.183110
31      d       top  300  50.971626

c. Which recipe, on average, is the best? recipe b

In [121]:
cake_melt.groupby("recipe").mean()

score
recipe           
a       63.922201
b       76.736074
c       75.874748
d       62.864844

d. Which oven temperature, on average, produces the best results? 275

In [122]:
cake_melt.groupby("temp").mean()

score
temp           
225   71.306022
250   66.577437
275   74.886754
300   66.627655

e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [135]:
(cake_melt.groupby(["recipe", "postition", "temp"]).mean()).idxmax()

score    (b, bottom, 300)
dtype: object